File: backend/app/scheduler.py
import asyncio
import logging
import uuid
from dataclasses import dataclass, field
from datetime import datetime
from enum import Enum
from typing import TYPE_CHECKING, Literal, Optional

if TYPE_CHECKING:
    from .mqtt_client import MQTTClient


logger = logging.getLogger(__name__)


class MessageStatus(str, Enum):
    PENDING = "Pending"
    ACTIVE = "Active"
    COMPLETED = "Completed"


Priority = Literal["normal", "high"]
PRIORITY_RANK = {"normal": 0, "high": 1}


@dataclass
class Message:
    id: uuid.UUID
    message: str
    created_at: datetime
    status: MessageStatus
    display_duration: int
    target_display_count: int
    display_count: int
    last_displayed_at: Optional[datetime]
    priority: Priority = "normal"

    def to_dict(self) -> dict:
        return {
            "id": str(self.id),
            "message": self.message,
            "createdAt": self.created_at.isoformat(),
            "status": self.status.value,
            "displayDuration": self.display_duration,
            "targetDisplayCount": self.target_display_count,
            "displayCount": self.display_count,
            "lastDisplayedAt": self.last_displayed_at.isoformat() if self.last_displayed_at else None,
            "priority": self.priority,
        }


class MessageStore:
    def __init__(self) -> None:
        self._messages: dict[uuid.UUID, Message] = {}
        self._lock = asyncio.Lock()

    async def add(self, message: Message) -> None:
        async with self._lock:
            self._messages[message.id] = message

    async def get(self, message_id: uuid.UUID) -> Optional[Message]:
        async with self._lock:
            return self._messages.get(message_id)

    async def list_active(self) -> list[Message]:
        async with self._lock:
            return [m for m in self._messages.values() if m.status != MessageStatus.COMPLETED]

    async def list_all(self) -> list[Message]:
        async with self._lock:
            return list(self._messages.values())

    async def mark_completed(self, message_id: uuid.UUID) -> bool:
        async with self._lock:
            m = self._messages.get(message_id)
            if m is None:
                return False
            m.status = MessageStatus.COMPLETED
            return True

    async def update_fields(self, message_id: uuid.UUID, **kwargs) -> None:
        async with self._lock:
            m = self._messages.get(message_id)
            if m is None:
                return
            for k, v in kwargs.items():
                setattr(m, k, v)


class Scheduler:
    def __init__(
        self,
        mqtt: "MQTTClient",
        publish_topic: str,
        default_display_duration: int,
        default_target_display_count: int,
        idle_message: str,
        idle_mode: str,
        idle_publish_interval: int,
    ) -> None:
        self._mqtt = mqtt
        self._publish_topic = publish_topic
        self._default_display_duration = default_display_duration
        self._default_target_display_count = default_target_display_count
        self._idle_message = idle_message
        self._idle_mode = idle_mode
        self._idle_publish_interval = idle_publish_interval

        self._store = MessageStore()
        self._current: Optional[Message] = None
        self._lock = asyncio.Lock()
        self._wakeup = asyncio.Event()
        self._task: Optional[asyncio.Task] = None

        self._subscribers: set[asyncio.Queue] = set()

    # ---- Public API (spec §14) ----

    async def add_message(
        self,
        text: str,
        target_display_count: Optional[int] = None,
        display_duration: Optional[int] = None,
        priority: Priority = "normal",
    ) -> uuid.UUID:
        if not text or not text.strip():
            raise ValueError("text must be non-empty")

        target = target_display_count if target_display_count is not None else self._default_target_display_count
        duration = display_duration if display_duration is not None else self._default_display_duration

        if target <= 0:
            raise ValueError("target_display_count must be > 0")
        if duration <= 0:
            raise ValueError("display_duration must be > 0")
        if priority not in ("normal", "high"):
            raise ValueError("priority must be 'normal' or 'high'")

        msg = Message(
            id=uuid.uuid4(),
            message=text,
            created_at=datetime.now(),
            status=MessageStatus.PENDING,
            display_duration=duration,
            target_display_count=target,
            display_count=0,
            last_displayed_at=None,
            priority=priority,
        )
        await self._store.add(msg)
        self._wakeup.set()
        self._notify({"type": "queue", "message": msg.to_dict()})
        return msg.id

    async def remove_message(self, message_id: uuid.UUID) -> bool:
        ok = await self._store.mark_completed(message_id)
        if ok:
            self._notify({"type": "queue", "removed": str(message_id)})
        return ok

    def get_active_messages(self) -> list[Message]:
        return [m for m in self._store._messages.values() if m.status != MessageStatus.COMPLETED]

    def get_all_messages(self) -> list[Message]:
        return list(self._store._messages.values())

    def get_current_message(self) -> Optional[Message]:
        return self._current

    def select_next_message(self) -> Optional[Message]:
        candidates = [
            m for m in self._store._messages.values()
            if m.status != MessageStatus.COMPLETED
        ]
        if not candidates:
            return None
        return min(
            candidates,
            key=lambda m: (
                -PRIORITY_RANK[m.priority],
                m.display_count,
                m.created_at,
            ),
        )

    def state(self) -> str:
        if self._current is not None:
            return "Active"
        if self.get_active_messages():
            return "Active"
        return "Idle"

    def high_priority_count(self) -> int:
        return sum(1 for m in self.get_active_messages() if m.priority == "high")

    # ---- Lifecycle ----

    async def start(self) -> None:
        if self._task is None:
            self._task = asyncio.create_task(self._run_loop())

    async def stop(self) -> None:
        if self._task is not None:
            self._task.cancel()
            try:
                await self._task
            except asyncio.CancelledError:
                pass
            self._task = None

    async def _run_loop(self) -> None:
        while True:
            try:
                await self.scheduler_tick()
            except asyncio.CancelledError:
                raise
            except Exception:
                logger.exception("scheduler_tick failed")
                await asyncio.sleep(1)

    # ---- Tick (spec §14) ----

    async def scheduler_tick(self) -> None:
        next_msg = self.select_next_message()
        if next_msg is None:
            await self._handle_idle()
            return

        self._current = next_msg
        self._notify({"type": "current", "message": next_msg.to_dict()})

        try:
            await self._mqtt.publish(self._publish_topic, next_msg.message, qos=0)
        except RuntimeError as e:
            logger.warning("Publish failed; will retry next tick: %s", e)
            self._current = None
            self._notify({"type": "current", "message": None})
            await asyncio.sleep(1)
            return
        except Exception:
            logger.exception("Publish raised; will retry next tick")
            self._current = None
            self._notify({"type": "current", "message": None})
            await asyncio.sleep(1)
            return

        next_msg.last_displayed_at = datetime.now()
        self._notify({"type": "current", "message": next_msg.to_dict()})

        # Full displayDuration — no early wake (spec §9).
        await asyncio.sleep(next_msg.display_duration)

        async with self._lock:
            next_msg.display_count += 1
            if next_msg.display_count >= 1 and next_msg.status == MessageStatus.PENDING:
                next_msg.status = MessageStatus.ACTIVE
            completed = next_msg.display_count >= next_msg.target_display_count
            if completed:
                next_msg.status = MessageStatus.COMPLETED
                self._current = None

        self._notify({"type": "current", "message": self._current.to_dict() if self._current else None})
        self._notify({"type": "queue", "message": next_msg.to_dict()})

    async def _handle_idle(self) -> None:
        if self._current is not None:
            self._current = None
            self._notify({"type": "current", "message": None})

        if self._idle_mode == "keep":
            try:
                await asyncio.wait_for(self._wakeup.wait(), timeout=1.0)
                self._wakeup.clear()
            except asyncio.TimeoutError:
                pass
            return

        # "publish" mode
        try:
            await self._mqtt.publish(self._publish_topic, self._idle_message, qos=0)
        except (RuntimeError, Exception):
            pass

        try:
            await asyncio.wait_for(
                self._wakeup.wait(),
                timeout=self._idle_publish_interval,
            )
            self._wakeup.clear()
        except asyncio.TimeoutError:
            pass

    # ---- Subscribers (SSE) ----

    def subscribe_queue(self) -> asyncio.Queue:
        q: asyncio.Queue = asyncio.Queue(maxsize=500)
        self._subscribers.add(q)
        return q

    def unsubscribe_queue(self, q: asyncio.Queue) -> None:
        self._subscribers.discard(q)

    def _notify(self, event: dict) -> None:
        for q in list(self._subscribers):
            try:
                q.put_nowait(event)
            except asyncio.QueueFull:
                try:
                    q.get_nowait()
                except asyncio.QueueEmpty:
                    pass
                try:
                    q.put_nowait(event)
                except asyncio.QueueFull:
                    pass

File: backend-go/internal/scheduler/scheduler.go
// Package scheduler implements the queueing, MQTT publishing and idle
// behaviour of the split-flap display backend.
package scheduler

import (
	"context"
	"encoding/json"
	"errors"
	"log"
	"sort"
	"strings"
	"sync"
	"sync/atomic"
	"time"

	"splitflap-web/internal/models"
	"splitflap-web/internal/mqttclient"
	"splitflap-web/internal/queue"
	"splitflap-web/internal/store"
)

// ErrInvalidInput is returned by AddMessage for invalid arguments.
var ErrInvalidInput = errors.New("invalid input")

// Scheduler is the scheduling core.
type Scheduler struct {
	mq                    *mqttclient.Client
	publishTopic          string
	defaultDisplayDur     int
	defaultTargetCount    int
	idleMessage           string
	idleMode              string
	idleInterval          int

	store                 *store.Store
	mu                    sync.Mutex
	current               *models.Message
	wakeup                chan struct{}

	// SSE subscriber set.
	subsMu sync.Mutex
	subs   map[*queue.Queue]struct{}

	// History ring buffer, max 50, prepend.
	histMu sync.Mutex
	hist   []models.HistoryEntry

	running atomic.Bool
	cancel   context.CancelFunc
	done     chan struct{}
}

// New constructs a new scheduler. The scheduler is not started; call Start.
func New(mq *mqttclient.Client,
	publishTopic string,
	defaultDisplayDur, defaultTargetCount int,
	idleMessage, idleMode string,
	idleInterval int,
) *Scheduler {
	return &Scheduler{
		mq:                 mq,
		publishTopic:       publishTopic,
		defaultDisplayDur:  defaultDisplayDur,
		defaultTargetCount: defaultTargetCount,
		idleMessage:        idleMessage,
		idleMode:           idleMode,
		idleInterval:       idleInterval,
		store:              store.New(),
		wakeup:             make(chan struct{}, 1),
		subs:               make(map[*queue.Queue]struct{}),
		hist:               make([]models.HistoryEntry, 0, 50),
		done:               make(chan struct{}),
	}
}

// --- Public accessors ------------------------------------------------------

// GetActiveMessages returns the list of non-Completed messages.
func (s *Scheduler) GetActiveMessages() []*models.Message {
	return s.store.ListActive()
}

// GetAllMessages returns every stored message.
func (s *Scheduler) GetAllMessages() []*models.Message {
	return s.store.ListAll()
}

// GetHistory returns a copy of the history ring buffer.
func (s *Scheduler) GetHistory() []models.HistoryEntry {
	s.histMu.Lock()
	defer s.histMu.Unlock()
	out := make([]models.HistoryEntry, len(s.hist))
	copy(out, s.hist)
	return out
}

// GetCurrentMessage returns the currently-displayed message, or nil.
func (s *Scheduler) GetCurrentMessage() *models.Message {
	s.mu.Lock()
	defer s.mu.Unlock()
	return s.current
}

// State returns "Active" if a message is current or any message is active;
// otherwise "Idle".
func (s *Scheduler) State() string {
	s.mu.Lock()
	hasCurrent := s.current != nil
	s.mu.Unlock()
	if hasCurrent {
		return "Active"
	}
	if len(s.store.ListActive()) > 0 {
		return "Active"
	}
	return "Idle"
}

// HighPriorityCount returns the number of non-Completed high-priority
// messages.
func (s *Scheduler) HighPriorityCount() int {
	count := 0
	for _, m := range s.store.ListActive() {
		if m.Priority == models.PriorityHigh {
			count++
		}
	}
	return count
}

// QueueSnapshot returns the active messages sorted in queue order.
func (s *Scheduler) QueueSnapshot() []models.MessageDTO {
	active := s.store.ListActive()
	dtos := make([]models.MessageDTO, 0, len(active))
	for _, m := range active {
		dtos = append(dtos, m.ToDTO())
	}
	// Sort by queue order: high priority first, then display_count asc,
	// then created_at asc.
	sort.SliceStable(dtos, func(i, j int) bool {
		ri := prioritySortRank(models.Priority(dtos[i].Priority))
		rj := prioritySortRank(models.Priority(dtos[j].Priority))
		if ri != rj {
			return ri < rj
		}
		if dtos[i].DisplayCount != dtos[j].DisplayCount {
			return dtos[i].DisplayCount < dtos[j].DisplayCount
		}
		return dtos[i].CreatedAt < dtos[j].CreatedAt
	})
	return dtos
}

// prioritySortRank maps high -> 0, normal -> 1 (used for ascending sort).
func prioritySortRank(p models.Priority) int {
	if p == models.PriorityHigh {
		return 0
	}
	return 1
}

// HistorySnapshot returns DTOs of all history entries (already-sorted
// format: most recent first).
func (s *Scheduler) HistorySnapshot() []models.HistoryEntry {
	return s.GetHistory()
}

// --- Message ops ----------------------------------------------------------

// AddMessage validates the input, creates a new Message in the store,
// records history, wakes the scheduler if idle, and notifies SSE
// subscribers. It returns the new message ID and a validation error.
func (s *Scheduler) AddMessage(text string, targetDisplayCount, displayDuration *int, priority *models.Priority, user string) (string, error) {
	if strings.TrimSpace(text) == "" {
		return "", validationError("text must be non-empty")
	}
	tdc := s.defaultTargetCount
	if targetDisplayCount != nil {
		if *targetDisplayCount <= 0 {
			return "", validationError("target_display_count must be > 0")
		}
		tdc = *targetDisplayCount
	}
	dd := s.defaultDisplayDur
	if displayDuration != nil {
		if *displayDuration <= 0 {
			return "", validationError("display_duration must be > 0")
		}
		dd = *displayDuration
	}
	pr := models.PriorityNormal
	if priority != nil {
		if *priority != models.PriorityNormal && *priority != models.PriorityHigh {
			return "", validationError("priority must be 'normal' or 'high'")
		}
		pr = *priority
	}

	m := models.NewMessage(strings.TrimSpace(text), tdc, dd, pr, user)
	s.store.Add(m)
	s.recordHistory(m)
	s.signalWakeup()

	subs := s.collectSubscribers()
	qData, _ := json.Marshal(struct {
		Type     string                `json:"type"`
		Messages []models.MessageDTO   `json:"messages"`
	}{Type: "queue", Messages: s.QueueSnapshot()})
	hData, _ := json.Marshal(struct {
		Type     string                `json:"type"`
		Messages []models.HistoryEntry `json:"messages"`
	}{Type: "history", Messages: s.HistorySnapshot()})
	for _, q := range subs {
		q.Push(queue.Event{Name: "queue", Data: qData})
		q.Push(queue.Event{Name: "history", Data: hData})
	}

	return m.ID, nil
}

// RemoveMessage marks the message as Completed. Returns true if found.
func (s *Scheduler) RemoveMessage(id string) bool {
	ok := s.store.MarkCompleted(id)
	if !ok {
		return false
	}
	subs := s.collectSubscribers()
	qData, _ := json.Marshal(struct {
		Type     string              `json:"type"`
		Messages []models.MessageDTO `json:"messages"`
	}{Type: "queue", Messages: s.QueueSnapshot()})
	for _, q := range subs {
		q.Push(queue.Event{Name: "queue", Data: qData})
	}
	return true
}

// --- SSE subscriber system -----------------------------------------------

// SubscribeQueue registers a queue to receive SSE lifecycle events.
// The queue is seeded with current/queue/history snapshots.
func (s *Scheduler) SubscribeQueue() *queue.Queue {
	q := queue.New(500)
	s.mu.Lock()
	curr := s.currentSnapshot()
	s.mu.Unlock()
	currData, _ := marshalCurrent(curr)
	qData, _ := json.Marshal(struct {
		Type     string              `json:"type"`
		Messages []models.MessageDTO `json:"messages"`
	}{Type: "queue", Messages: s.QueueSnapshot()})
	hData, _ := json.Marshal(struct {
		Type     string                `json:"type"`
		Messages []models.HistoryEntry `json:"messages"`
	}{Type: "history", Messages: s.HistorySnapshot()})
	q.Push(queue.Event{Name: "current", Data: currData})
	q.Push(queue.Event{Name: "queue", Data: qData})
	q.Push(queue.Event{Name: "history", Data: hData})

	s.subsMu.Lock()
	s.subs[q] = struct{}{}
	s.subsMu.Unlock()
	return q
}

// UnsubscribeQueue removes the given queue from the subscriber set.
func (s *Scheduler) UnsubscribeQueue(q *queue.Queue) {
	s.subsMu.Lock()
	delete(s.subs, q)
	s.subsMu.Unlock()
}

// currentSnapshot returns a MessageDTO or nil marshalled-ready pointer
// along with its raw JSON for the "current" event.
func (s *Scheduler) currentSnapshot() (ptr *models.MessageDTO) {
	if s.current == nil {
		return nil
	}
	dto := s.current.ToDTO()
	return &dto
}

func marshalCurrent(curr *models.MessageDTO) ([]byte, error) {
	return json.Marshal(struct {
		Type    string               `json:"type"`
		Message *models.MessageDTO  `json:"message"`
	}{Type: "current", Message: curr})
}

func (s *Scheduler) collectSubscribers() []*queue.Queue {
	s.subsMu.Lock()
	defer s.subsMu.Unlock()
	out := make([]*queue.Queue, 0, len(s.subs))
	for q := range s.subs {
		out = append(out, q)
	}
	return out
}

func (s *Scheduler) notify(event queue.Event) {
	for _, q := range s.collectSubscribers() {
		q.Push(event)
	}
}

// --- History -------------------------------------------------------------

func (s *Scheduler) recordHistory(m *models.Message) {
	e := models.NewHistoryEntry(m)
	s.histMu.Lock()
	defer s.histMu.Unlock()
	// Prepend.
	s.hist = append([]models.HistoryEntry{e}, s.hist...)
	if len(s.hist) > 50 {
		s.hist = s.hist[:50]
	}
}

// --- Wake up -------------------------------------------------------------

func (s *Scheduler) signalWakeup() {
	select {
	case s.wakeup <- struct{}{}:
	default:
	}
}

// drainWakeup consumes any pending wakeup signal. Must be called only when
// we are sure no wait will be performed immediately afterwards.
func (s *Scheduler) drainWakeup() {
	for {
		select {
		case <-s.wakeup:
		default:
			return
		}
	}
}

// waitWakeup blocks for up to d. Returns true if woken by a new message.
// The provided context lets the runner cancel the wait on shutdown.
func (s *Scheduler) waitWakeup(ctx context.Context, d time.Duration) bool {
	t := time.NewTimer(d)
	defer t.Stop()
	select {
	case <-s.wakeup:
		// Clear remaining signals (buffer size 1 so we are done).
		s.drainWakeup()
		return true
	case <-t.C:
		return false
	case <-ctx.Done():
		return false
	}
}

// --- Selection ------------------------------------------------------------

// SelectNextMessage returns the next message to display according to the
// sort key (-priority_rank, display_count, created_at). Returns nil if no
// message is selectable.
func (s *Scheduler) SelectNextMessage() *models.Message {
	active := s.store.ListActive()
	if len(active) == 0 {
		return nil
	}
	sort.SliceStable(active, func(i, j int) bool {
		ai, aj := active[i], active[j]
		ri := -models.PriorityRank(ai.Priority)
		rj := -models.PriorityRank(aj.Priority)
		if ri != rj {
			return ri < rj
		}
		if ai.DisplayCount != aj.DisplayCount {
			return ai.DisplayCount < aj.DisplayCount
		}
		return ai.CreatedAt.Before(aj.CreatedAt)
	})
	return active[0]
}

// --- Tick ---------------------------------------------------------------

// schedulerTick runs a single iteration of the scheduling loop.
func (s *Scheduler) schedulerTick(ctx context.Context) error {
	m := s.SelectNextMessage()
	if m == nil {
		s.handleIdle(ctx)
		return nil
	}

	s.mu.Lock()
	s.current = m
	s.mu.Unlock()

	currData, _ := json.Marshal(struct {
		Type    string              `json:"type"`
		Message *models.MessageDTO  `json:"message"`
	}{Type: "current", Message: ptrToDTO(m)})
	s.notify(queue.Event{Name: "current", Data: currData})

	// Attempt publish.
	if err := s.mq.Publish(s.publishTopic, m.Message, 0); err != nil {
		log.Printf("scheduler: publish failed: %v", err)
		s.mu.Lock()
		s.current = nil
		s.mu.Unlock()
		nullCurr, _ := json.Marshal(struct {
			Type    string             `json:"type"`
			Message *models.MessageDTO `json:"message"`
		}{Type: "current", Message: nil})
		s.notify(queue.Event{Name: "current", Data: nullCurr})
		// backoff
		select {
		case <-time.After(1 * time.Second):
		case <-ctx.Done():
			return ctx.Err()
		}
		return nil
	}

	// Successful publish: record last_displayed_at, increment count,
	// update status.
	now := time.Now()
	s.store.Update(m.ID, func(mm *models.Message) {
		mm.LastDisplayedAt = &now
		mm.DisplayCount = mm.DisplayCount + 1
		if mm.DisplayCount >= 1 && mm.Status == models.StatusPending {
			mm.Status = models.StatusActive
		}
	})
	// After update, the in-memory `m` is the same stored pointer.
	m.LastDisplayedAt = &now
	m.DisplayCount = m.DisplayCount + 1
	if m.DisplayCount >= 1 && m.Status == models.StatusPending {
		m.Status = models.StatusActive
	}

	// Emit current (with lastDisplayedAt).
	currData2, _ := json.Marshal(struct {
		Type    string              `json:"type"`
		Message *models.MessageDTO  `json:"message"`
	}{Type: "current", Message: ptrToDTO(m)})
	s.notify(queue.Event{Name: "current", Data: currData2})

	// Emit queue snapshot BEFORE the dwell sleep.
	qData, _ := json.Marshal(struct {
		Type     string              `json:"type"`
		Messages []models.MessageDTO `json:"messages"`
	}{Type: "queue", Messages: s.QueueSnapshot()})
	s.notify(queue.Event{Name: "queue", Data: qData})

	// Dwell: sleep full display_duration. NO early wake.
	select {
	case <-time.After(time.Duration(m.DisplayDuration) * time.Second):
	case <-ctx.Done():
		return ctx.Err()
	}

	// After dwell: check completion.
	if m.DisplayCount >= m.TargetDisplayCount {
		s.store.MarkCompleted(m.ID)
		m.Status = models.StatusCompleted
		s.mu.Lock()
		s.current = nil
		s.mu.Unlock()
		qDataDone, _ := json.Marshal(struct {
			Type     string              `json:"type"`
			Messages []models.MessageDTO `json:"messages"`
		}{Type: "queue", Messages: s.QueueSnapshot()})
		s.notify(queue.Event{Name: "queue", Data: qDataDone})
	} else {
		// Refresh history (next tick will reselect based on counts).
		// per spec not needed here.
	}

	// Notify current (after dwell, possibly null).
	s.mu.Lock()
	var curr *models.MessageDTO
	if s.current != nil {
		dto := s.current.ToDTO()
		curr = &dto
	}
	s.mu.Unlock()
	currData3, _ := json.Marshal(struct {
		Type    string             `json:"type"`
		Message *models.MessageDTO `json:"message"`
	}{Type: "current", Message: curr})
	s.notify(queue.Event{Name: "current", Data: currData3})
	return nil
}

// handleIdle is invoked when select_next_message returns null.
func (s *Scheduler) handleIdle(ctx context.Context) {
	s.mu.Lock()
	if s.current != nil {
		s.current = nil
		s.mu.Unlock()
		nullCurr, _ := json.Marshal(struct {
			Type    string             `json:"type"`
			Message *models.MessageDTO `json:"message"`
		}{Type: "current", Message: nil})
		s.notify(queue.Event{Name: "current", Data: nullCurr})
	} else {
		s.mu.Unlock()
	}

	switch s.idleMode {
	case "keep":
		s.waitWakeup(ctx, 1*time.Second)
	case "publish", "":
		if err := s.mq.Publish(s.publishTopic, s.idleMessage, 0); err != nil {
			// Silently ignore idle publish errors.
			_ = err
		}
		s.waitWakeup(ctx, time.Duration(s.idleInterval)*time.Second)
	default:
		// Unknown idle_mode: behave as "keep" to avoid busy-looping.
		s.waitWakeup(ctx, 1*time.Second)
	}
}

// --- Run loop ------------------------------------------------------------

// Start launches the scheduler loop goroutine. Idempotent.
func (s *Scheduler) Start(ctx context.Context) {
	if !s.running.CompareAndSwap(false, true) {
		return
	}
	ctx, cancel := context.WithCancel(ctx)
	s.cancel = cancel
	go func() {
		defer close(s.done)
		for {
			if err := s.schedulerTick(ctx); err != nil {
				if errors.Is(err, context.Canceled) {
					return
				}
				log.Printf("scheduler: tick error: %v", err)
				select {
				case <-time.After(1 * time.Second):
				case <-ctx.Done():
					return
				}
				continue
			}
			// Check for shutdown between ticks: handleIdle returns nil
			// even if the ctx was canceled.
			select {
			case <-ctx.Done():
				return
			default:
			}
		}
	}()
}

// Stop cancels the scheduler loop and waits for it to finish.
func (s *Scheduler) Stop() {
	if s.cancel != nil {
		s.cancel()
	}
	<-s.done
}

// --- Helpers -------------------------------------------------------------

func ptrToDTO(m *models.Message) *models.MessageDTO {
	if m == nil {
		return nil
	}
	dto := m.ToDTO()
	return &dto
}

func validationError(msg string) error {
	return &ValidationError{Msg: msg}
}

// ValidationError is returned for invalid user input. Its message is
// intended to be sent verbatim to the HTTP client.
type ValidationError struct{ Msg string }

func (e *ValidationError) Error() string { return e.Msg }